### PR TITLE
Add source-code editor and simple integration

### PR DESCRIPTION
**Pull request by smacker**
screen shot 2018-05-21 at 15 54 00

HOC returns only props due need to support different layouts and it's an application responsibility, not a lib.

### CODE REVIEW

File: src/withUASTEditor.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    onCursorChanged(cursorPos) {
+      const pos = this.posIndex.get({
+        Line: cursorPos.line + 1,
+        Col: cursorPos.ch + 1
+      });
+      if (!pos) {
+        return;
+      }
+      const { uast, lastHighlighted } = this.state;
+      const newUast = highlightNodeById(uast, pos.id, lastHighlighted);
```
Comment (carlosms) [inline src/withUASTEditor.js]: I think I've found a bug here. If you click on the same word, the node gets un-highlighted.

One possible solution:

```diff
 export function highlightNodeById(uast, id, prevId) {
   const node = uast[id];
   const newUast = {
     ...uast,
     [id]: {
       ...node,
       highlighted: true
     }
   };
-  if (prevId !== null) {
+  if (prevId !== null && prevId !== id) {
     newUast[prevId] = {
       ...newUast[prevId],
       highlighted: false
     };
   }

```

File: src/withUASTEditor.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          makePositionIndexHook(this.posIndex)
+        ),
+        // control highlighted node
+        lastHighlighted: null,
+        // position of hovered node
+        pos: null
+      };
+    }
+
+    onNodeHover(id, prevId) {
```
Comment (carlosms) [inline src/withUASTEditor.js]: In the demo I'm seeing something strange.

The node hover effect is persistent, when the mouse exists the UAST tree you still get the highlighted text in the editor. But then you click in the text, and a different UAST node gets highlighted. I think there should be a way to clear state.pos when the mouse exists the UAST tree.

File: src/Editor.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  render() {
+    const [props, childProps] = splitProps(this.props, Editor);
+    const { code, languageMode, onChange } = props;
+
+    const options = {
+      mode: languageMode,
+      lineNumbers: true,
+      theme: 'solarized light'
```
Comment (carlosms) [inline src/Editor.js]: Maybe the theme should be open to be configured via props, with this one as default.

File: src/withUASTEditor.js (hunk 2 above)
Comment (smacker) [reply src/withUASTEditor.js]: fixed

File: src/withUASTEditor.js (hunk 1 above)
Comment (smacker) [reply src/withUASTEditor.js]: fixed

**Comment by bzz** [comment]
@carlosms seems like feedback has been addressed and it's time for another pass.

@smacker seems like a rebase is in order here

**Comment by smacker** [comment]
rebased. Commits are squashed to simplify rebase.

Demo moved to styleguide:
screen shot 2018-05-22 at 11 57 37

Added md file for Editor component.

**Pull request merged**